### PR TITLE
Open last active cluster after switching workspaces

### LINKED ISSUE #827: Dashboard doesn't refresh after changing workspace
**Describe the bug**
When changing from one workspace to another the dashboard is "blank" (there is no current cluster shown) 

**To Reproduce**
1. With two workspaces, click on workspace widget (bottom left) and change to the other workspace
2. Observe that the workspace is shown, with no active cluster(even if there was an active cluster when the workspace was last selected)
3. Click on a cluster icon to cause the refresh

**Expected behavior**
When toggling between workspaces the dashboard should show previous snapshot of the workspace
**Screenshots**

wsRefresh

**Environment (please complete the following information):**
- Lens Version: 3.6.0-rc.1
- OS: OSX
- Installation method (e.g. snap or AppImage in Linux): dmg

### PR DESCRIPTION
**Pull request by aleksfront**
Based on https://github.com/lensapp/lens/pull/1054/files

Saves last active cluster for each workspace and opens it when workspaces switched.

switching workspaces

Fixes #827

### CODE REVIEW

File: src/common/workspace-store.ts [hunk 1]
```diff
@@ -184,6 +187,13 @@ export class WorkspaceStore extends BaseStore<WorkspaceStoreModel> {
     clusterStore.removeByWorkspaceId(id)
   }
 
+  @action
+  setLastActiveClusterId(clusterId: ClusterId, workspaceId = this.currentWorkspaceId) {
+    if (clusterId != null) {
```
Comment (Nokel81) [inline src/common/workspace-store.ts]: This possibility should be marked in the signature

File: src/renderer/components/+workspaces/workspace-menu.tsx [hunk 2]
```diff
@@ -17,6 +20,16 @@ interface Props extends Partial<MenuProps> {
 export class WorkspaceMenu extends React.Component<Props> {
   @observable menuVisible = false;
 
+  activateWorkspace = (id: WorkspaceId) => {
+    const clusterId = workspaceStore.getById(id).lastActiveClusterId;
```
Comment (Nokel81) [inline src/renderer/components/+workspaces/workspace-menu.tsx:24]: What happens when there is no `lastActiveClusterId`?

File: src/renderer/components/+workspaces/workspace-menu.tsx [hunk 3]
```diff
@@ -17,6 +20,16 @@ interface Props extends Partial<MenuProps> {
...
+    workspaceStore.setActive(id);
+    if (clusterId) {
+      navigate(clusterViewURL({ params: { clusterId } }));
+    } else {
+      navigate(landingURL());
+    }
```
Comment (Nokel81) [inline src/renderer/components/+workspaces/workspace-menu.tsx:30]: Why is this code different from the code in `workspaces.tsx`?

File: src/renderer/components/cluster-manager/clusters-menu.tsx [hunk 4]
```diff
@@ -57,7 +57,11 @@ export class ClustersMenu extends React.Component<Props> {
       menu.append(new MenuItem({
         label: _i18n._(t`Disconnect`),
         click: async () => {
+          const lastActiveClusterId = workspaceStore.currentWorkspace.lastActiveClusterId;
           if (clusterStore.isActive(cluster.id)) {
+            if (lastActiveClusterId === cluster.id) {
+              workspaceStore.setLastActiveClusterId("");
```
Comment (Nokel81) [inline src/renderer/components/cluster-manager/clusters-menu.tsx]: This is very weird, `""` is not a valid `ClusterId` (so some function might throw if this gets out). There should be a better way to represent this (`undefined` maybe given that `lastActiveClusterId` is optional.

File: src/renderer/components/+workspaces/workspace-menu.tsx (hunk 2 above)
Comment (aleksfront) [reply src/renderer/components/+workspaces/workspace-menu.tsx:24]: If `lastActiveCluster` id became `undefined` then user will be redirected to Landing page following code below.

File: src/renderer/components/+workspaces/workspace-menu.tsx (hunk 3 above)
Comment (aleksfront) [reply src/renderer/components/+workspaces/workspace-menu.tsx:30]: When you hit on workspace link from Workspaces page, there's no redirect to dashboard - user will stay on same page and see that `current` indicator will jump to clicked workspace.

File: src/renderer/components/cluster-manager/clusters-menu.tsx (hunk 4 above)
Comment (aleksfront) [reply src/renderer/components/cluster-manager/clusters-menu.tsx]: Good catch! Refactored this part. A lot of code removed!

**Comment by Nokel81** [comment]
Rerequest my review when you are ready for it

**Comment by aleksfront** [comment]
> Rerequest my review when you are ready for it

It's ready, except for failed tests, need to fix them. Please take a second look.

File: src/common/workspace-store.ts [hunk 5]
```diff
@@ -7,10 +7,12 @@ import { broadcastMessage } from "../common/ipc";
 import logger from "../main/logger";
 
 export type WorkspaceId = string;
+export type ClusterId = string;
 
 export interface WorkspaceStoreModel {
+  workspaces: WorkspaceModel[];
   currentWorkspace?: WorkspaceId;
-  workspaces: WorkspaceModel[]
+  lastActiveClusterId?: ClusterId;
```
Comment (Nokel81) [inline src/common/workspace-store.ts]: Wait, does this mean that we only track the last active cluster once for the whole application (and per workspace)?

File: src/common/workspace-store.ts [hunk 6]
```diff
@@ -7,10 +7,12 @@ import { broadcastMessage } from "../common/ipc";
 import logger from "../main/logger";
 
 export type WorkspaceId = string;
+export type ClusterId = string;
```
Comment (Nokel81) [inline src/common/workspace-store.ts]: This should be imported from some place else

File: src/common/workspace-store.ts (hunk 5 above)
Comment (Nokel81) [reply src/common/workspace-store.ts]: Or are we only saving the last active cluster globally (still seems odd)?

Comment (aleksfront) [reply src/common/workspace-store.ts]: We're tracking last active cluster for each workspace. This field belongs to `Workspace` class.

File: src/common/workspace-store.ts (hunk 6 above)
Comment (aleksfront) [reply src/common/workspace-store.ts]: Fixed

File: src/common/workspace-store.ts (hunk 5 above)
Comment (Nokel81) [reply src/common/workspace-store.ts]: See my other (sniped) comment

Comment (aleksfront) [reply src/common/workspace-store.ts]: You mean this one? https://github.com/lensapp/lens/pull/1444#discussion_r526988672
Isn't it about same question about clarification of where `lastActiveClusterId` field is located?

Comment (Nokel81) [reply src/common/workspace-store.ts]: It is, I noticed that I had asked about the one located on `WorkspaceStoreModel` which seems strange to me. I would have though that it should be been on `WorskaceModel`

Comment (aleksfront) [reply src/common/workspace-store.ts]: Oh, that's true. Moved into `WorkspaceModel`.

**Review by Nokel81 — APPROVED** [review]
Looks good!

File: src/common/workspace-store.ts (hunk 6 above)
Comment (ixrock) [reply src/common/workspace-store.ts]: It's already defined in `cluster-store.ts` as:
```
export type ClusterId = string;
```
So just `import type {ClusterId} from "../cluster-store.ts"`

File: src/common/cluster-store.ts [hunk 7]
```diff
@@ -11,9 +11,10 @@ import { appEventBus } from "./event-bus"
 import { dumpConfigYaml } from "./kube-helpers";
 import { saveToAppFiles } from "./utils/saveToAppFiles";
 import { KubeConfig } from "@kubernetes/client-node";
+import { subscribeToBroadcast, unsubscribeAllFromBroadcast } from "./ipc";
 import _ from "lodash";
 import move from "array-move";
-import { subscribeToBroadcast, unsubscribeAllFromBroadcast } from "./ipc";
+import type { WorkspaceId } from "./workspace-store";
```
Comment (ixrock) [inline src/common/cluster-store.ts:17]: You already importing it as normal module on the top => so no need to for extra import?

File: src/renderer/components/+workspaces/workspace-menu.tsx (hunk 2 above)
Comment (ixrock) [reply src/renderer/components/+workspaces/workspace-menu.tsx:24]: Maybe it's better to select first cluster in workspace and show it instead landing page? If no clusters -> landing?

File: src/common/cluster-store.ts (hunk 7 above)
Comment (aleksfront) [reply src/common/cluster-store.ts:17]: If not imported separately as a `type`, webpack warnings will appear in compilation time.

**Pull request merged**